docs: Initial gatsby-plugin-image reference guide

ascorbic opened the pull request:
First pass at a reference document for the image plugin. It needs more on the sizing options, but we ened to make some decisions there. It also needs docs for the component-only props, such as `loading`

[Rendered version](https://github.com/gatsbyjs/gatsby/blob/docs/image-plugin-refs/docs/docs/reference/built-in-components/gatsby-plugin-image.md)

[ch22668]

laurieontech reviewed (commented):
This is great. Think calling out the most common options is super helpful. Left some comments and suggestions.

ascorbic commented on docs/docs/reference/built-in-components/gatsby-plugin-image.md:
```diff
@@ -0,0 +1,113 @@
+
+## Image options
+
+The API for both components is almost the same. The difference is how you pass the properties. When using `StaticImage`, they are passed as props to the component, whereas for the `GatsbyImage` component they are passed to the `gatsbyImageData` GraphQL resolver. There is a small change in the values too: in the `StaticImage` component, props such as layout and placeholder take a string, while the ...
```
Changed to "The options..."

meganesu commented on docs/docs/reference/built-in-components/gatsby-plugin-image.md:
````diff
@@ -0,0 +1,129 @@
+}
+```
+
+#### Constrained
````
Throwing this out there as an idea: What do you think about putting these values into a table instead of paragraphs? Trying to make this page easier to scan.

Something like: 

| Name | Component prop | Resolver prop | Description |
| --- | --- | --- | --- |
| Constrained | `"constrained"` | `CONSTRAINED` | This is the default layout. It displays... |

I'm imagining something similar to this vibe: https://material-ui.com/api/radio/

meganesu replied:
Or actually, maybe that makes more sense for the table of contents at the beginning of the doc (~ line 19), since the descriptions aren't so long.

Then it could be:

| Prop | Value | Default Value | Description |
| --- | --- | --- | --- |
| `layout` | constrained or fixed or full width | constrained | ... |

Thoughts?

meganesu commented on docs/docs/reference/built-in-components/gatsby-plugin-image.md:
```diff
@@ -0,0 +1,129 @@
+
+Use this for images that are always displayed at the full width of the screen, such as banners or hero images. Like the constrained layout, this resizes to fit the container. However it is not restricted to a maximum size, so will grow to fill the container however large it is, maintaining its aspect ratio. It generates several smaller image sizes for different screen breakpoints, so that the bro ...
+
+### Sizing options
```
For consistency, I'd love it if all the prop names could be at the same heading level. (vs. right now there's a difference between the two sizing options and the other three.)

How would you feel about getting rid of this "Sizing options" level and just having two separate h3s for `width / height` and `aspectRatio`?

meganesu replied:
If we go down that route, the note on line 73 ("Size props are optional...") could maybe go in the `width / height` section? Or combined with the paragraph on line 81 ("If you set just one of these...")?

meganesu commented on docs/docs/reference/built-in-components/gatsby-plugin-image.md:
```diff
@@ -0,0 +1,129 @@
+
+#### Dominant color
+
+_Component prop: `"dominantColor"`. Resolver prop: `DOMINANT_COLOR`_
```
If we end up using a table for the `layout` values, we should use a similar table here for the `placeholder` values.

meganesu commented on docs/docs/reference/built-in-components/gatsby-plugin-image.md:
```diff
@@ -0,0 +1,129 @@
+- **`StaticImage`:** Use this if the image is the same every time the component is used. _Examples: site logo, index page hero image_
+- **`GatsbyImage`:** Use this if the image is passed-in to the component as a prop. _Examples: Blog post hero image, author avatar_
+
+## Image options
```
Are these options all optional? (Sounds silly when I write it out like that...)

It might be good to add a quick intro sentence to give this section some context. I don't love this wording, but something like: "If you'd like to override the default behavior of `StaticImage` and `GatsbyImage`, you can pass in one or more additional options."

laurieontech replied:
Nothing is required except for alt. Which is not explicitly listed but in all the examples, so we should probably adjust that.

meganesu commented on docs/docs/reference/built-in-components/gatsby-plugin-image.md:
```diff
@@ -0,0 +1,129 @@
+title: Gatsby Image plugin
+---
+
+This guide will show you how to configure your images, including choosing layouts, placeholders and image processing options. If you would like to learn how to set up the image plugins and use images on your site, see [the how-to guide](/docs/how-to/images-and-media/using-gatsby-plugin-image).
```
> If you would like to learn how to set up the image plugins and use images on your site, see [the how-to guide](/docs/how-to/images-and-media/using-gatsby-plugin-image).

I'd move this sentence down after you list out the components. So that it's like "Here are the two components you can use. If you want an example, go to the how-to guide."

meganesu replied:
Maybe in this introduction we could put a similar callout as in the other doc, sending people to the Conceptal Guide about image optimizations.

meganesu commented on docs/docs/reference/built-in-components/gatsby-plugin-image.md:
```diff
@@ -0,0 +1,129 @@
+
+### Placeholder
+
+**Prop name: `placeholder`**
```
This section feels like another good candidate for a video (similar to the layout one) to show what the different options look like.

meganesu reviewed (changes requested):
Left some ideas about formatting. Let me know if you have any thoughts or questions! Overall, this is really clearly written. Nicely done! 🎉

ascorbic commented:
@meganesu I've added tables as suggested, and also added all of the other options to the table at the end. I don't know if you htink they should be in another doc instead

ascorbic commented:
@meganesu I've now fleshed this out and added docs on all of the component props too. I don't know whether you think it should be split into a separate API doc or not. I'm inclined to think it's ok as it is. What do you think?

LekoArts commented on docs/docs/reference/built-in-components/gatsby-plugin-image.md:
```diff
@@ -0,0 +1,183 @@
+| ----------------- | --------------------------------------------------------------------------- | ------------- | ----------------------------------------------------------------------------------------------------------------------------------------------- |
+| `alt` (Required)  | `string`                                                                    |               | Alternative text, passed to the `<img>` tag. Required for accessibility.                                                                        |
+| `as`              | `ElementType`                                                               | `"div"`       | The HTML element used for the outer wrapper.                                                                                                    |
+| `loading`         | `"eager" \| "lazy"`                                                         | `"lazy"`      | Loading behavior for the image. You should set this to `"eager"` for above-the-fold images to ensure they start loading before React hydration. |
```
I think the escaping of the pipe doesn't work with MDX on .com, so while GitHub is fine, the website will show the backslash. I've used the mathematical divisor in the past -- see "Documentation" of this PR https://github.com/gatsbyjs/gatsby/pull/28485

ascorbic replied:
Thanks! It doesn't work in Typora either. Great fix.

LekoArts commented on docs/docs/reference/built-in-components/gatsby-plugin-image.md:
```diff
@@ -0,0 +1,183 @@
+
+For `fullWidth` images you don't specify width or height as it resizes to fit the screen width. Passing `aspectRatio` will crop the image if needed, and the height will scale according to the width of the screen. For example, if you set the aspectRatio to `16/9` then when the image is displayed full width on a screen that is 1024px wide, the image will be 576 pixels high.
+
+> There are several advanced options that you can pass to control the cropping and resizing behavior. For more details, see the advanced options reference.
```
Anchor link here?

LekoArts commented on docs/docs/reference/built-in-components/gatsby-plugin-image.md:
```diff
@@ -0,0 +1,183 @@
+
+For `fixed` and `constrained` images, you can also optionally pass either a `width` or `height`, and it will use that to calculate the other dimension. For example, if you pass `width={800} aspectRatio={4/3}` then `height` will be set to the width divided by the aspect ratio: so `600`. Passing `1` as the aspectRatio will crop the image to a square. If you don't pass a width or height then it will ...
+
+For `fullWidth` images you don't specify width or height as it resizes to fit the screen width. Passing `aspectRatio` will crop the image if needed, and the height will scale according to the width of the screen. For example, if you set the aspectRatio to `16/9` then when the image is displayed full width on a screen that is 1024px wide, the image will be 576 pixels high.
```
Just as an idea: How about using 1280px and 720px? Maybe HD and it's 16:9 ratio is more common

ascorbic replied:
@LekoArts  Great idea. Thanks!

LekoArts reviewed (commented):
Awesome doc!

meganesu commented on docs/docs/reference/built-in-components/gatsby-plugin-image.md:
```diff
@@ -0,0 +1,185 @@
+
+### `formats`
+
+_Default component prop value: `["auto", "webp"]`. Default resolver prop value: `[AUTO, WEBP]`_
```
Was this line left here intentionally? Might be able to get rid of it, since it's in the table up above. (But I also don't mind having the duplication here - my callout is mostly just that this is the only option that still has a default listed this way.)

meganesu commented on docs/docs/reference/built-in-components/gatsby-plugin-image.md:
```diff
@@ -0,0 +1,185 @@
+
+| Option                                  | Default                                                              | Description                                                                                                                                                                                                                                                                                  ...
+| --------------------------------------- | -------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ...
+| [`layout`](#layout)                     | `"constrained"` / `CONSTRAINED`                                      | Determines the size of the image and its resizing behavior.                                                                                                                                                                                                                                  ...
+| [`width`/`height`](#widthheight)        | Source image size                                                    | Change the size of the image.                                                                                                                                                                                                                                                                ...
+| [`aspectRatio`](#aspectratio)           | Source image aspect ratio                                            | Force a specific ratio between the image's width and height.                                                                                                                                                                                                                                 ...
+| [`placeholder`](#placeholder)           | `"dominantColor"`/`DOMINANT_COLOR`                                   | Choose the style of temporary image shown while the full image loads.                                                                                                                                                                                                                        ...
+| [`formats`](#formats)                   | `["auto","webp"]`/`[AUTO,WEBP]`                                      | File formats of the images generated.                                                                                                                                                                                                                                                        ...
+| [`transformOptions`](#transformoptions) | `{fit: "cover", cropFocus: "attention"}`                             | Options to pass to sharp to control cropping and other image manipulations.                                                                                                                                                                                                                  ...
```
Is there a reason you included these ones here again? What would you think about changing this section from "all options" to "additional sharp options" or something like that?

ascorbic replied:
I did it so there was a place to see them all, and make it clear that there's nothing special about the others, just that they're more common.

meganesu replied:
Cool, that makes sense to me!

meganesu reviewed (dismissed):
Left a few more comments. I also think it's fine to leave this all in one doc.

This is looking beautiful! Great job! ✨👏

meganesu commented on docs/docs/reference/built-in-components/gatsby-plugin-image.md:
```diff
@@ -0,0 +1,185 @@
+| Prop             | Type     | Description                                                                                           |
+| ---------------- | -------- | ----------------------------------------------------------------------------------------------------- |
+| `src` (Required) | `string` | Source image, processed at build time. Can be a path relative to the source file, or an absolute URL. |
+
```
````suggestion

#### Restrictions on using `StaticImage`

The images are loaded and processed at build time, so there are restrictions on how you pass props to the component. The values need to be statically-analyzed at build time, which means you can't pass them as props from outside the component, or use the results of function calls, for example. You can either use static values, or variables within the component's local scope. See the following examples:

This does not work:

```js
// ⚠️ Doesn't work
export function Logo({ logo }) {
  // You can't use a prop passed into the parent component
  return <StaticImage src={logo}>
}
```

...and nor does this:

```js
// ⚠️ Doesn't work
export function Dino() {
    // Props can't come from function calls
    const width = getTheWidthFromSomewhere();
    return <StaticImage src="trex.png" width={width}>
}
```

You can use variables and expressions if they're in the scope of the file, e.g.:

```js
// OK
export function Dino()  {
    // Local variables are fine
    const width = 300
    return <StaticImage src="trex.png" width={width}>
}
```

```js
// Also OK
// A variable in the same file is fine.
const width = 300
export function Dino()  {
    // This works because the value can be statically-analyzed
    const height = width * 16 / 9
    return <StaticImage src="trex.png" width={width} height={height}>
}
```

If you find yourself wishing you could use a prop for the image `src` then it's likely that you should be using a dynamic image.

````

meganesu commented on docs/docs/reference/built-in-components/gatsby-plugin-image.md:
````diff
@@ -0,0 +1,234 @@
+export function Logo({ logo }) {
+  // You can't use a prop passed into the parent component
+  return <StaticImage src={logo}>
+}
+```
+
+...and nor does this:
+
+```js
+// ⚠️ Doesn't work
+export function Dino() {
+    // Props can't come from function calls
+    const width = getTheWidthFromSomewhere();
+    return <StaticImage src="trex.png" width={width}>
+}
+```
+
+You can use variables and expressions if they're in the scope of the file, e.g.:
+
+```js
+// OK
+export function Dino()  {
+    // Local variables are fine
+    const width = 300
+    return <StaticImage src="trex.png" width={width}>
+}
+```
+
+```js
+// Also OK
+// A variable in the same file is fine.
+const width = 300
+export function Dino()  {
+    // This works because the value can be statically-analyzed
+    const height = width * 16 / 9
+    return <StaticImage src="trex.png" width={width} height={height}>
+}
+```
+
+If you find yourself wishing you could use a prop for the image `src` then it's likely that you should be using a dynamic image.
````
```suggestion
**Note:** There are a few technical restrictions to the way you can pass props into StaticImage. For more information, refer to the Reference Guide: [Gatsby Image plugin](/docs/reference/built-in-components/gatsby-plugin-image#restrictions-on-using-staticimage). If you find yourself wishing you could use a prop for the image `src` then it's likely that you should be using a dynamic image.
```

meganesu commented:
Ugh I'm a goober and just realized I switched the restrictions and the callout, so now they're in the wrong docs 🤦‍♀️  Too many tabs open at once. I'll open another PR to fix it.
